src/xdsl/parser.py

```diff
@@ -608,27 +605,48 @@ def parse_optional_attribute(self,
             symbol_name = self.parse_alpha_num(skip_white_space=False)
             return FlatSymbolRefAttr.from_str(symbol_name)
 
-        parsed = self.parse_optional_char("!")
-        if parsed is None:
-            return None
+        def parse_integer_type():
+            self.parse_char("!", skip_white_space=skip_white_space)
+            self.parse_char("i", skip_white_space=False)
+            val = self.parse_int_literal(skip_white_space=False)
+            return IntegerType.from_width(val)
+
+        if int_type := self.try_parse(parse_integer_type):
+            return int_type
+
+        return None
+
+    def parse_optional_attribute(self,
+                                 skip_white_space: bool = True
+                                 ) -> Attribute | None:
+        if self.source == self.Source.MLIR:
```

Comment:
Maybe some comments here would help for the future. This should be MLIR costum parsing, the next is xDSL costum parsing, right?